src/parse.jl

```diff
@@ -101,3 +102,523 @@ function ZonedDateTime(str::AbstractString, format::AbstractString; locale::Abst
+        end
+        i = ii
+
+        # Convert to the Julia day-of-week used by `Dates.dayofweek`
+        # Equivalent to: `dow = (dow - 7) % 7 + 7`
+        day_of_week == 0 && (day_of_week = 7)
+
+        f = function (year::Integer)
+            date = Date(year, month, (week_of_month - 1) * 7 + 1)
+            if week_of_month < 5
+                step = Day(1)
+            else
+                date = lastdayofmonth(date)
+                step = Day(-1)
+            end
+            tonext(d -> dayofweek(d) == day_of_week, date; step=step, same=true, limit=7)
+        end
+
+        return f, i
+    else
+        val = tryparsenext_base10(str, i, len)
+        if val === nothing
+            return ParseNextError("Expected to find number of Julian days", str, i)
+        end
+        days, ii = val
+        if days < 0 || days > 365
+            # Note: On non-leap years day 365 will be the first day of the next year. This
+            # is only supported as the tzset(3) man page explicitly states that the
+            # zero-based Julia day is between 0 and 365.
+            return ParseNextError("Julian days outside of expected range [0, 365]", str, i, ii)
+        end
+        i = ii
+
+        # No prefix specifies the zero-based day of year between 0 and 365. February 29 is
+        # counted in leap years.
+        days += 1
+
+        f = (year::Integer) -> Date(year, 1) + Day(days - 1)
+        return f, i
+    end
+end
+
+"""
+    _parsesub_time(str, [i, len]) -> Union{Tuple{Integer, Integer}, Exception}
+
+Parses a section of a string as a time of the form `hh[:mm[:ss]]`. Primarily this function
+is used to parse daylight saving transition times as outlined in tzset(3).
+"""
+function _parsesub_time(
+    str::AbstractString,
+    i::Integer=firstindex(str),
+    len::Integer=lastindex(str);
+    name::AbstractString="time",
+)
+    if i > len
+        return ParseNextError("Expected $name and instead found end of string", str, i)
+    end
+
+    # Require time does not start with a sign.
+    c, ii = iterate(str, i)::Tuple{Char, Int}
+    if c in ('+', '-')
+        return ParseNextError("$(uppercasefirst(name)) should not have a sign", str, i)
+    end
+
+    return _parsesub_offset(str, i, len; name=name)
+end
+
+"""
+    _parsesub_tz(str, [i, len]) -> Union{Tuple{TimeZone, Integer}, Exception}
+
+Parse a direct representation of a time zone as specified by the tzset(3) man page.
+
+# Examples
+```jldoctest
+julia> first(_parsesub_tz("EST+5"))
+EST (UTC-5)
+
+julia> first(_parsesub_tz("NZST-12:00:00NZDT-13:00:00,M10.1.0,M3.3.0"))
+NZST/NZDT (UTC+12/UTC+13)
+"""
+function _parsesub_tz(
+    str::AbstractString,
+    i::Integer=firstindex(str),
+    len::Integer=lastindex(str),
+)
+    # An empty or unset TZ environmental variable defaults to UTC
+    if len == 0
+        return FixedTimeZone("UTC"), i
+    end
+
+    x = _parsesub_tzabbr(str, i, len)
+    if x isa Tuple
```

Comment:
```suggestion
    x isa Tuple ? (stdname, i = x) : return x
```

Might be easier to read here and below. Feel free to resolve if you don't agree

Reply:
I'm not sure that's better. I want to do some refactoring of this code later and I expect this pattern will be hidden by a macro